[LBSE] Add RenderSVGResourceContainer

rwlbuis opened the pull request:
#### a9d6d01a743c404d3c502d0bf663b87088cc3772
<pre>
[LBSE] Add RenderSVGResourceContainer
<a href="https://bugs.webkit.org/show_bug.cgi?id=261880">https://bugs.webkit.org/show_bug.cgi?id=261880</a>

Reviewed by Nikolas Zimmermann.

Add RenderSVGResourceContainer for LBSE which knows how to handle RenderLayer
and will act for future LBSE specific SVG resources.
Start keeping track of RenderSVGResourceContainer in ShadowRoot and adjust
RenderObject to be able to distinguish between LBSE and legacy SVG resources.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::addSVGResource):
(WebCore::TreeScope::removeSVGResource):
(WebCore::TreeScope::svgResourceById const):
* Source/WebCore/dom/TreeScope.h:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isLegacySVGResourceContainer const):
* Source/WebCore/rendering/svg/RenderSVGHiddenContainer.cpp:
(WebCore::RenderSVGHiddenContainer::RenderSVGHiddenContainer):
* Source/WebCore/rendering/svg/RenderSVGHiddenContainer.h:
(WebCore::RenderObject::isLegacySVGResourceContainer const):
* Source/WebCore/rendering/svg/RenderSVGHiddenContainer.h:
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp: Added.
(WebCore::RenderSVGResourceContainer::RenderSVGResourceContainer):
(WebCore::RenderSVGResourceContainer::willBeDestroyed):
(WebCore::RenderSVGResourceContainer::styleDidChange):
(WebCore::RenderSVGResourceContainer::idChanged):
(WebCore::RenderSVGResourceContainer::registerResource):
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.h: Added.
* Source/WebCore/rendering/svg/SVGResourcesCycleSolver.cpp:
(WebCore::SVGResourcesCycleSolver::resourceContainsCycles):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceContainer.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
* Source/WebCore/svg/SVGDefsElement.cpp:
(WebCore::SVGDefsElement::createElementRenderer):
* Source/WebCore/svg/SVGGElement.cpp:
(WebCore::SVGGElement::createElementRenderer):
* Source/WebCore/svg/SVGSymbolElement.cpp:
(WebCore::SVGSymbolElement::createElementRenderer):

Canonical link: <a href="https://commits.webkit.org/269522@main">https://commits.webkit.org/269522@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/40cf4b9994e5547b4a108a75d0f1f012d1bc3188

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22814 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/679 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23901 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24722 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21117 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23071 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1583 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23340 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22013 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23054 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/300 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19785 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25575 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20665 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/26873 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/20660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20911 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24730 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/354 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18170 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/279 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5434 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/430 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/384 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/dom/TreeScope.cpp:
```diff
@@ -583,7 +583,8 @@ struct SVGResourcesMap {
     MemoryCompactRobinHoodHashMap<AtomString, WeakHashSet<SVGElement, WeakPtrImplWithEventTargetData>> pendingResources;
     MemoryCompactRobinHoodHashMap<AtomString, WeakHashSet<SVGElement, WeakPtrImplWithEventTargetData>> pendingResourcesForRemoval;
-    MemoryCompactRobinHoodHashMap<AtomString, LegacyRenderSVGResourceContainer*> resources;
+//    MemoryCompactRobinHoodHashMap<AtomString, RenderSVGResourceContainer*> resources;
```
We don't land commented code.

rwlbuis replied:
Yeah, that is a good rule :) Though this PR is first to get some feedback.

nikolaszimmermann commented on Source/WebCore/dom/TreeScope.cpp:
```diff
@@ -593,29 +594,38 @@ SVGResourcesMap& TreeScope::svgResourcesMap() const
+        return;
+
+    // Replaces resource if already present, to handle potential id changes
+    // svgResourcesMap().resources.set(id, &resource);
```
Why is it commented out? :-)

rwlbuis replied:
Does not seem needed to comment out indeed, reinstated.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
```diff
@@ -0,0 +1,248 @@
+
+void RenderSVGResourceContainer::markAllClientsForInvalidation(InvalidationMode mode)
+{
+    // FIXME: Style invalidation should either be a pre-layout task or this function
```
I think we should try hard to get these things right this time.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
```diff
@@ -0,0 +1,248 @@
+
+    bool needsLayout = mode == LayoutAndBoundariesInvalidation;
+    bool markForInvalidation = mode != ParentOnlyInvalidation;
+    auto* root = SVGRenderSupport::findTreeRootObject(*this);
```
That uses the non-Legacy* variant. I suppose?

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
```diff
@@ -0,0 +1,248 @@
+    markAllClientLayersForInvalidation();
+}
+
+void RenderSVGResourceContainer::markAllClientLayersForInvalidation()
```
I am not sure we should actually add all of that invalidation handling in the beginning, but rather get e.g. simple clipping running, and then revisit the whole topic regarding invalidation, when we know for sure what clients we have to serve: RenderLayers, or RenderObjects, or both, etc.

WDYT?

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
```diff
@@ -0,0 +1,248 @@
+{
+    ASSERT(!m_clients.isEmptyIgnoringNullReferences());
+
+    switch (mode) {
```
I doubt we still need 'LayoutAndBoundaries'/'BoundariesInvalidation' - best to not start at all with this kind of code.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
```diff
@@ -0,0 +1,248 @@
+        auto* renderer = client->renderer();
+        if (!renderer)
+            continue;
+        SVGResourcesCache::clientStyleChanged(*renderer, StyleDifference::Layout, nullptr, renderer->style());
```
Also very dubious -- what exactly does that do? We should avoid these question in the new resources system. IMHO we should avoid any use of SVGResourceCache::client* in LBSE, from the beginning.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
```diff
@@ -0,0 +1,248 @@
+    }
+}
+
+bool RenderSVGResourceContainer::shouldTransformOnTextPainting(const RenderElement& renderer, AffineTransform& resourceTransform)
```
I'd rather add a smaller skeleton, and avoid all these special code inherited from legacy SVG, let's see if we need this if we get to text + gradients...

rwlbuis replied:
Sure, removed.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
```diff
@@ -0,0 +1,248 @@
+#endif
+}
+
+// FIXME: This does not belong here.
```
Right, nuke it :-)

rwlbuis replied:
Removed.

nikolaszimmermann reviewed (commented):
This needs further discussions. For LBSE I tried hard to not add anything from legacy that is poorly understood/unclear: the whole resources concept is a mess, with SVGResourcesCache::clientFooChanged() / all the markFor*Invalidation code (unclear when it should run; not during layout, as we do right now, for sure).

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (C) Research In Motion Limited 2010. All rights reserved.
```
And Igalia 2023 copyright?

rwlbuis replied:
Done.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceContainer.h:
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (C) Research In Motion Limited 2010. All rights reserved.
```
Igalia copyright?

rwlbuis replied:
Done.

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGResourcesCycleSolver.cpp:
```diff
@@ -46,7 +46,7 @@ bool SVGResourcesCycleSolver::resourceContainsCycles(LegacyRenderSVGResourceCont
     RenderObject* node = &resource;
     while (node) {
-        if (node != &resource && node->isSVGResourceContainer()) {
+        if (node != &resource && node->isLegacySVGResourceContainer()) {
```
We need to double-check cycle solving. But out of scope for this patch...

rwlbuis replied:
Agreed :)

nikolaszimmermann reviewed (approved):
r=me

rwlbuis committed:
[LBSE] Add RenderSVGResourceContainer
https://bugs.webkit.org/show_bug.cgi?id=261880

Reviewed by Nikolas Zimmermann.

Add RenderSVGResourceContainer for LBSE which knows how to handle RenderLayer
and will act for future LBSE specific SVG resources.
Start keeping track of RenderSVGResourceContainer in ShadowRoot and adjust
RenderObject to be able to distinguish between LBSE and legacy SVG resources.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::addSVGResource):
(WebCore::TreeScope::removeSVGResource):
(WebCore::TreeScope::svgResourceById const):
* Source/WebCore/dom/TreeScope.h:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isLegacySVGResourceContainer const):
* Source/WebCore/rendering/svg/RenderSVGHiddenContainer.cpp:
(WebCore::RenderSVGHiddenContainer::RenderSVGHiddenContainer):
* Source/WebCore/rendering/svg/RenderSVGHiddenContainer.h:
(WebCore::RenderObject::isLegacySVGResourceContainer const):
* Source/WebCore/rendering/svg/RenderSVGHiddenContainer.h:
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp: Added.
(WebCore::RenderSVGResourceContainer::RenderSVGResourceContainer):
(WebCore::RenderSVGResourceContainer::willBeDestroyed):
(WebCore::RenderSVGResourceContainer::styleDidChange):
(WebCore::RenderSVGResourceContainer::idChanged):
(WebCore::RenderSVGResourceContainer::registerResource):
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.h: Added.
* Source/WebCore/rendering/svg/SVGResourcesCycleSolver.cpp:
(WebCore::SVGResourcesCycleSolver::resourceContainsCycles):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceContainer.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
* Source/WebCore/svg/SVGDefsElement.cpp:
(WebCore::SVGDefsElement::createElementRenderer):
* Source/WebCore/svg/SVGGElement.cpp:
(WebCore::SVGGElement::createElementRenderer):
* Source/WebCore/svg/SVGSymbolElement.cpp:
(WebCore::SVGSymbolElement::createElementRenderer):

Canonical link: https://commits.webkit.org/269522@main
